bpo-34616: add flags to allow top-level-await

Carreau opened the pull request:
Based on a work with @1st1 during PyCon 2019; and can be seen used in https://github.com/ipython/ipython/pull/11713 (which would dramatically reduce thee complexity and shave an extra few hundreds lines of IPython if we were supporting 3.8+ only).

<!-- issue-number: [bpo-34616](https://bugs.python.org/issue34616) -->
https://bugs.python.org/issue34616
<!-- /issue-number -->

Commit:
WIP async repl implementation

This now works:

    import asyncio
    import types

    print('here')

    L = {
        'asyncio': asyncio,
        'a': 'something'
    }

    code = compile('''
    async with asyncio.Lock():
        a = await asyncio.sleep(1, result=42)
    ''', '', 'single', 0x2000)

    f = types.FunctionType(code, L)

    print('before', L)
    asyncio.run(f())
    print('after', L)

Commit:
Add option to `compile()` to accept top-level await features.

In some narrow but rate case; it is useful to be allow to compile code
with top-level await features; for example when writing a REPL.
This attempt to allow that.

1st1 commented on asyncex:
```diff
@@ -0,0 +1,20 @@
+import asyncio
```
This file shouldn't be part of the PR.  We'll need to add a couple of unit tests though.

1st1 commented on trycomp:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
This file shouldn't be part of the PR.

1st1 commented on Python/compile.c:
```diff
@@ -4759,12 +4765,20 @@ compiler_visit_expr1(struct compiler *c, expr_ty e)
+            c->u->u_ste->ste_coroutine = 1;
+        }
+
+        if (!(c->c_flags->cf_flags & PyCF_ALLOW_TOP_LEVEL_AWAIT)) {
```
This check should be the top-level if; then nest the other two checks inside.

1st1 commented on Python/compile.c:
```diff
@@ -4759,12 +4765,20 @@ compiler_visit_expr1(struct compiler *c, expr_ty e)
-        if (c->u->u_scope_type != COMPILER_SCOPE_ASYNC_FUNCTION &&
-                c->u->u_scope_type != COMPILER_SCOPE_COMPREHENSION)
-            return compiler_error(c, "'await' outside async function");
+            c->u->u_ste->ste_coroutine = 1;
```
I'm not sure if this is actually necessary, try to remove this line and see if the tests pass (and add tests :)).

Carreau replied:
... well it actually is or I get 
```
unhandled exception during asyncio.run() shutdown
task: <Task finished name='Task-2' coro=<<async_generator_athrow without __name__>()> exception=RuntimeError("can't send non-None value to a just-started coroutine")>
RuntimeError: can't send non-None value to a just-started coroutine
```

Not quite sure I understand exactly why...

1st1 commented on Python/compile.c:
```diff
@@ -4759,12 +4763,17 @@ compiler_visit_expr1(struct compiler *c, expr_ty e)
+                return compiler_error(c, "'await' outside async function");
+            }
+        }
+        c->u->u_ste->ste_coroutine = 1;
```
I'm still not sure if this line is needed, but if it is, it should only be set when `(c->c_flags->cf_flags & PyCF_ALLOW_TOP_LEVEL_AWAIT)` is true.

Carreau commented on Python/compile.c:
```diff
@@ -5699,6 +5708,10 @@ compute_code_flags(struct compiler *c)
     flags |= (c->c_flags->cf_flags & PyCF_MASK);
 
+    if ((c->c_flags->cf_flags & PyCF_ALLOW_TOP_LEVEL_AWAIT) && c->u->u_ste->ste_coroutine) {
+        flags |= CO_COROUTINE;
```
This thing has some weird effect and turn async_generators into coroutine;  I'm going to change that to `!ste->ste_generator`

1st1 replied:
Please fix to fit 79 characters

Carreau replied:
Done.

Commit:
remove some unnecessary already existing coroutine flags

Commit:
add actual necessary ste_couroutines

Commit:
stry new line

Commit:
Reformat some code

tirkarthi commented on Doc/library/functions.rst:
```diff
@@ -257,6 +257,12 @@ are always available.  They are listed here in alphabetical order.
+   The optional argument *flags* also control wether the compiled source is
+   allowed to contain top-level ``await``, ``async for`` and ``async with``.
+   When the bit ``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` is set, the the return code
+   object has ``CO_COROUTINE`` set in ``co_code``, and can be interacitvely
```
```suggestion
   object has ``CO_COROUTINE`` set in ``co_code``, and can be interactively
```

tirkarthi commented on Doc/library/functions.rst:
```diff
@@ -257,6 +257,12 @@ are always available.  They are listed here in alphabetical order.
    can be found as the :attr:`~__future__._Feature.compiler_flag` attribute on
    the :class:`~__future__._Feature` instance in the :mod:`__future__` module.
 
+   The optional argument *flags* also control wether the compiled source is
```
```suggestion
   The optional argument *flags* also controls whether the compiled source is
```

tirkarthi commented on Lib/test/test_builtin.py:
```diff
@@ -358,6 +361,32 @@ def f(): """doc"""
 
+    def test_compile_top_level_await(self):
+        for mode in ('single', 'exec'):
+            for num,code_sample in enumerate(['''await sleep(0)''',
```
Minor nit : num is not used in test so `enumerate` seems to be not needed . Perhaps the list used in enumerate can be defined as a variable outside of for loop and just iterated?

Carreau replied:
Good point, I used it to debug which sample(s) were wrong and did not removed it, will do.

tirkarthi commented on Lib/test/test_builtin.py:
```diff
@@ -358,6 +361,32 @@ def f(): """doc"""
+                                       pass''']):
+                source = dedent(code_sample)
+                with self.assertRaises(SyntaxError,
+                                  msg='source={!r} mode={!r})'.format(source, mode)):
```
Align msg with `SyntaxError` ? PEP 8 enforces 80 character limit per line but I will leave it to reviewer on the same.

Carreau replied:
I've done that and used `itertools.product` to remove one level of indent. I'm still slightly over.

tirkarthi replied:
Perhaps have `msg` as a variable outside of context manager.

```python
msg = 'source={!r} mode={!r})'.format(source, mode)

with self.assertRaises(SyntaxError, msg):
    compile(source, '?' , mode)
```

or even better master has f-string debugging merged into master where `f"{source=}"` expands into `f"source={source!r}"`. So you can rebase onto master and write below : 

```python
with self.assertRaises(SyntaxError, msg=f'{source=} {mode=})'): # 75 characters
    compile(source, '?' , mode)
```

Carreau replied:
Yep good idea. Done that and rebased on master.

Carreau committed:
Apply grammar corrections from code review

thanks to tirkarthi

Co-Authored-By: Xtreak <tir.karthi@gmail.com>

Carreau committed:
Cleanup test suite code formatting.

Use itertools product to remove one level of nesting, and make things a
tiny bit more readable.

Carreau committed:
document an extra test

Carreau committed:
better formatting

Carreau committed:
use debug-f-string

1st1 commented on Doc/library/functions.rst:
```diff
@@ -257,6 +257,12 @@ are always available.  They are listed here in alphabetical order.
+   allowed to contain top-level ``await``, ``async for`` and ``async with``.
+   When the bit ``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` is set, the the return code
+   object has ``CO_COROUTINE`` set in ``co_code``, and can be interactively
+   executed via ``yield from eval(code_object)``.
```
```suggestion
   executed via ``await eval(code_object)``.
```

1st1 commented on Doc/library/functions.rst:
```diff
@@ -290,6 +296,10 @@ are always available.  They are listed here in alphabetical order.
       in *source*.
 
+   .. versionadded:: 3.8
+      *flags* can now be used to accept source that would contain a top-level
```
``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` can now be passed in *flags* to enable support
for top-level await, async for, and async with.

Carreau replied:
Done.

1st1 commented on Lib/test/test_builtin.py:
```diff
@@ -358,6 +362,43 @@ def f(): """doc"""
+        co = compile(dedent("""async def ticker():
+                for i in range(10):
+                    yield i
+                    await asyncio.sleep(0)"""), '?', 'exec', flags=ast.PyCF_ALLOW_TOP_LEVEL_AWAIT)
```
Please make sure all new lines fit the 79 characters limit.

Carreau replied:
Done.

1st1 commented on Python/compile.c:
```diff
@@ -4564,7 +4566,9 @@ compiler_async_with(struct compiler *c, stmt_ty s, int pos)
     assert(s->kind == AsyncWith_kind);
-    if (c->u->u_scope_type != COMPILER_SCOPE_ASYNC_FUNCTION) {
+    if (c->c_flags->cf_flags & PyCF_ALLOW_TOP_LEVEL_AWAIT){
+        c->u->u_ste->ste_coroutine = 1;
```
Have you checked if this `c->u->u_ste->ste_coroutine = 1;` line is needed?  Does it work without it?

Carreau replied:
yes it is necessary, otherwise: 

```
>>> async with acm() as delta:
>>>     count += delta

Traceback (most recent call last):
  File "minirepl.py", line 97, in f
    co = compile(inp, '', 'single', flags=PyCF_ALLOW_TOP_LEVEL_AWAIT)
  File "<string>", line 2
SyntaxError: 'async with' outside async function
```

1st1 replied:
OK, cool, let's keep it then :)

Carreau replied:
And there is be a test in `Lib/test/test_builtin.py` for that  in case you ask.

1st1 commented on Lib/test/test_builtin.py:
```diff
@@ -358,6 +362,43 @@ def f(): """doc"""
+            co = compile(source, '?', mode, flags=ast.PyCF_ALLOW_TOP_LEVEL_AWAIT)
+
+            self.assertEqual(co.co_flags & CO_COROUTINE, CO_COROUTINE,
+                             msg=f"{source=} {mode=}")
```
Please add a functional test, that wraps the created code object in a FunctionType and then manually iterates the coroutine (or with asyncio) and checks the result.

Carreau replied:
I've added 2 with `FunctionTypes` and `eval()`

Carreau committed:
Update Doc/library/functions.rst

Co-Authored-By: Yury Selivanov <yury@magic.io>

Commit:
code cleanup

Commit:
Redo-some functional testing.

Make the the top=level-await codeobjects can actually be put into t a
FunctionType and run, and/or put inside `await eval()` and do actually
run.

Commit:
Fix line lenght

Commit:
skip await-eval

Commit:
fix docs

Carreau commented on Lib/test/test_builtin.py:
```diff
@@ -358,6 +364,72 @@ def f(): """doc"""
+                globals_ = {'asyncio': asyncio, 'a':0, 'arange': arange}
+                asyncio.run(eval(co, globals_))
+                self.assertEqual(globals_['a'], 1)
+        except Exception:
```
I rewrote most of this, to also run via a `FunctionType` – though I believe creating FunctionTypes directly is not supposed to be done by usercode. and via `run(eval())`

Carreau commented on Doc/library/functions.rst:
```diff
@@ -257,6 +257,12 @@ are always available.  They are listed here in alphabetical order.
 
+   The optional argument *flags* also controls whether the compiled source is
+   allowed to contain top-level ``await``, ``async for`` and ``async with``.
+   When the bit ``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` is set, the the return code
```
Note to self. Double "the" here.

Carreau commented on Doc/library/functions.rst:
```diff
@@ -290,6 +296,10 @@ are always available.  They are listed here in alphabetical order.
 
+   .. versionadded:: 3.8
+      ``ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`` can now be passed in flags to enable
+      support for top-level await, async for, and async with.
```
note to self. Put `async ...` between backquotes.

Commit:
better documentation formatting

Commit:
restaure ev policy in finally
